safehttp/plugins/htmlinject/htmlinject.go

```diff
@@ -0,0 +1,194 @@
+	OnTag string
+	// WithAttributes is a filter applied on tags to decide whether to run the Rule:
+	// only tags with the given attributes key:value will be matched.
+	WithAttributes map[string]string
+	// AddAttributes is a list of strings to add to the HTML in place of attributes.
+	// All the given strings will be appended verbatim after the matched tag so they
+	// should be prefixed with a space.
+	AddAttributes []string
+	// AddNodes is a list of nodes to append immediately after the openin tag that matched.
+	// This means that for elements that have a matching closing tag the added node will be
+	// a child node, for self-closing tags it will be a sibling.
+	AddNodes []string
+}
+
+func (r Rule) String() string { return r.Name }
+
+// Config is a slice of Rules that are somehow related to each other.
+type Config []Rule
+
+// CSPNoncesDefault is the default config for CSP Nonces. The rewritten template
+// expects the CSPNonce Func to be available in the template to provide nonces.
+var CSPNoncesDefault = CSPNonces(`nonce="{{CSPNonce}}"`)
+
+// CSPNonces constructs a Config to add CSP nonces to a template. The given nonce
+// attribute will be automatically prefixed with the required empty space.
+func CSPNonces(nonceAttr string) Config {
+	nonceAttr = " " + nonceAttr
+	return Config{
+		Rule{
+			Name:          "Nonces for scripts",
+			OnTag:         "script",
+			AddAttributes: []string{nonceAttr},
+		},
+		Rule{
+			Name:           "Nonces for link as=script rel=preload",
+			OnTag:          "link",
+			WithAttributes: map[string]string{"rel": "preload", "as": "script"},
+			AddAttributes:  []string{nonceAttr},
+		},
+	}
+}
+
+// XSRFTokensDefault is the default config to add hidden inputs to forms to provide
+// an anti-XSRF token. The rewritten template expects the XSRFToken Func to be available
+// in the template to provide tokens and sets the name for the sent value to be "xsrftoken".
+var XSRFTokensDefault = XSRFTokens(`<input type="hidden" name="xsrf-token" value="{{XSRFToken}}">`)
+
+// XSRFTokens constructs a Config to add the given string as a child node to forms.
+func XSRFTokens(inputTag string) Config {
+	return Config{Rule{
+		Name:     "XSRFTokens on forms",
+		OnTag:    "form",
+		AddNodes: []string{inputTag}}}
+}
+
+// Transform rewrites the given template according to the given configs.
+// If the passed io.Rewriter has a `Size() int64` method it will be used to pre-allocate buffers.
+func Transform(src io.Reader, cfg ...Config) (tpl string, _ error) {
+	rw := rewriter{
+		rules:     map[string][]Rule{},
+		tokenizer: html.NewTokenizer(src),
+		out:       &strings.Builder{},
+	}
+	if sizer, ok := src.(interface{ Size() int64 }); ok {
+		rw.out.Grow(int(sizer.Size()))
+	}
+	for _, c := range cfg {
+		for _, r := range c {
+			rw.rules[r.OnTag] = append(rw.rules[r.OnTag], r)
+		}
+	}
+	if err := rw.rewrite(); err != nil {
+		return "", fmt.Errorf("transforming template: %v", err)
```

Comment:
The "transforming template" is obvious from the function call. If the caller needs to be reminded of what they're calling, they can add it themselves.